demo06_one_and_one.py

```diff
@@ -4,30 +4,18 @@
 from demo05_basic_defender import move as move_defender
 from demo04_basic_attacker import move as move_attacker
 
-def move(turn, game):
+def move(bot, state):
     # create a combined game state to collect the states for both bots
-    if game.state is None:
+    if state is None:
         # initialization
-        combined_state = {'attacker' : None, 'defender' : None}
-    else:
-        # keep a copy of the game state (the move_defender and move_attacker
-        # functions are going to overwrite game.state, so we need a copy
-        # here so that we can reset game.state before returning)
-        combined_state = game.state.copy()
+        state = {'attacker' : None, 'defender' : None}
 
-    if turn == 0:
-        # fake the game.state
-        game.state = combined_state['defender']
-        next_move = move_defender(turn, game)
-        # collect the defender state in our own dictionary
-        combined_state['defender'] = game.state
+    if bot.turn == 0:
+        # ignore returned state from defender, we store it in our
+        # state dictionary anyway
+        next_move, _ = move_defender(bot, state['defender'])
```

Comment:
I don’t see how you can ignore the returned state here (and if you can, we’d need an example where the returned state will be put back into the main state anyway).

Reply:
Well, I could if the other two bots would know about passing empty dictionaries. But, OK, I fixed that. And it still feels much nicer than before, and, more importantly, understandable

Reply:
If the bots knew about passing empty dictionaries, we could keep the old system and just drop the `copy`, no?

The reason we do the state explicitly is exactly to avoid these assumptions, isn’t it?